File: Section_2/temperature_converter.cpp
#include <iostream>
#include <functional>

std::function<std::pair<double, int>(double, bool)> temperatureConverter() {
    int count = 0;  // Enclosed variable

    return [count](double temp, bool toCelsius = true) mutable {
        count++;  // Modifying enclosed variable
        return toCelsius
            ? std::make_pair((temp - 32) * 5 / 9, count)  // Convert to Celsius
            : std::make_pair((temp * 9 / 5) + 32, count);  // Convert to Fahrenheit
    };
}

int main() {
    auto convertTemp = temperatureConverter();
    auto result1 = convertTemp(100, true);

    std::cout << "100 degrees F is " << result1.first << " C"
    << std::endl;

    auto result2 = convertTemp(0, false);
    std::cout << "0 degrees C is " << result2.first << " F"
    << std::endl;

    return 0;
}

File: Section_2/temperature_converter.py
def temperature_converter():
    count = 0  # Enclosed variable

    def convert(temp, to_celsius=True):
        nonlocal count  # Accessing the enclosing variable
        count += 1
        if to_celsius:
            return (temp - 32) * 5 / 9, count  # Convert to Celsius
        else:
            return (temp * 9 / 5) + 32, count  # Convert to Fahrenheit

    return convert

convert_temp = temperature_converter()
print(f"100 degrees F is {convert_temp(100, True)[0]} C")
print(f"0 degrees C is {convert_temp(0, False)[0]} F")
